### PR TITLE
Refresh dynamic instance metadata for NodeGetInfo calls

### PR DESCRIPTION
**Pull request by torredil**
#### What type of PR is this?

/kind feature

#### What is this PR about? / Why do we need it?

When Kubernetes periodically calls NodeGetInfo, the driver should return updated information about the maximum number of attachable volumes. This commit modifies the driver to refresh dynamic metadata (ENI count) when calculating volume limits, rather than using cached values from driver initialization.

#### How was this change tested?

Attaching one additional ENI after driver startup and observing driver logs / CSINode object.

- Node plugin logs:

```
I0409 15:36:55.336607       1 node.go:574] "NodeGetInfo: called" args=""
I0409 15:36:55.338301       1 metadata.go:93] "Updated metadata" numAttachedENIs=1 numBlockDeviceMappings=0
I0409 15:36:55.338335       1 node.go:781] "getVolumesLimit:" instanceType="c5.large"
I0409 15:36:55.338344       1 node.go:785] "getVolumesLimit:" availableAttachments=28
I0409 15:36:55.338352       1 node.go:790] "getVolumesLimit:" numBlockDevices=0
I0409 15:36:55.338358       1 node.go:792] "getVolumesLimit:" reservedVolumeAttachments=1
I0409 15:36:55.338459       1 node.go:806] "getVolumesLimit:" ENIs=1
I0409 15:36:55.338513       1 node.go:808] "getVolumesLimit:" reservedSlots=0
I0409 15:36:55.338525       1 node.go:601] "NodeGetInfo:" maxVolumesPerNode=26
I0409 15:37:01.930657       1 identity.go:61] "Probe: called" args=""
I0409 15:37:06.338244       1 node.go:574] "NodeGetInfo: called" args=""
I0409 15:37:06.341084       1 metadata.go:93] "Updated metadata" numAttachedENIs=2 numBlockDeviceMappings=0
I0409 15:37:06.341253       1 node.go:781] "getVolumesLimit:" instanceType="c5.large"
I0409 15:37:06.341347       1 node.go:785] "getVolumesLimit:" availableAttachments=28
I0409 15:37:06.341438       1 node.go:790] "getVolumesLimit:" numBlockDevices=0
I0409 15:37:06.341509       1 node.go:792] "getVolumesLimit:" reservedVolumeAttachments=1
I0409 15:37:06.341659       1 node.go:806] "getVolumesLimit:" ENIs=2
I0409 15:37:06.341749       1 node.go:808] "getVolumesLimit:" reservedSlots=0
I0409 15:37:06.341832       1 node.go:601] "NodeGetInfo:" maxVolumesPerNode=25
```

- CSINode object before additional ENI is attached:
```
Name:               i-036afc1e0189c56ad
Labels:             <none>
Annotations:        storage.alpha.kubernetes.io/migrated-plugins:
                      kubernetes.io/aws-ebs,kubernetes.io/azure-disk,kubernetes.io/azure-file,kubernetes.io/cinder,kubernetes.io/gce-pd,kubernetes.io/portworx-v...
CreationTimestamp:  Wed, 09 Apr 2025 14:57:07 +0000
Spec:
  Drivers:
    ebs.csi.aws.com:
      Node ID:  i-036afc1e0189c56ad
      Allocatables:
        Count:        26
      Topology Keys:  [kubernetes.io/os topology.ebs.csi.aws.com/zone topology.kubernetes.io/zone]
Events:               <none>
```

- CSINode object after additional ENI is attached:
```
Name:               i-036afc1e0189c56ad
Labels:             <none>
Annotations:        storage.alpha.kubernetes.io/migrated-plugins:
                      kubernetes.io/aws-ebs,kubernetes.io/azure-disk,kubernetes.io/azure-file,kubernetes.io/cinder,kubernetes.io/gce-pd,kubernetes.io/portworx-v...
CreationTimestamp:  Wed, 09 Apr 2025 14:57:07 +0000
Spec:
  Drivers:
    ebs.csi.aws.com:
      Node ID:  i-036afc1e0189c56ad
      Allocatables:
        Count:        25
      Topology Keys:  [kubernetes.io/os topology.ebs.csi.aws.com/zone topology.kubernetes.io/zone]
Events:               <none>
```

- CSI Driver manifest used for testing:
```
kubectl get csidriver -o yaml                          
apiVersion: v1
items:
- apiVersion: storage.k8s.io/v1
  kind: CSIDriver
  metadata:
    annotations:
      meta.helm.sh/release-name: aws-ebs-csi-driver
      meta.helm.sh/release-namespace: kube-system
    creationTimestamp: "2025-04-09T15:33:11Z"
    labels:
      app.kubernetes.io/component: csi-driver
      app.kubernetes.io/instance: aws-ebs-csi-driver
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: aws-ebs-csi-driver
      app.kubernetes.io/version: 1.41.0
      helm.sh/chart: aws-ebs-csi-driver-2.41.0
    name: ebs.csi.aws.com
    resourceVersion: "7647"
    uid: 152bcd6e-b3c7-41ec-9af5-dfe4898c2e9b
  spec:
    attachRequired: true
    fsGroupPolicy: File
    nodeAllocatableUpdatePeriodSeconds: 11
    podInfoOnMount: false
    requiresRepublish: false
    seLinuxMount: false
    storageCapacity: false
    volumeLifecycleModes:
    - Persistent
kind: List
metadata:
  resourceVersion: ""
```

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by torredil** [comment]
@ElijahQuinones reminded us that IMDS [block device mapping](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/block-device-mapping-concepts.html) information is set at instance launch time and is immutable. Since querying that endpoint will not return updated information, I've removed the call from `UpdateMetadata()`.

**Comment by ElijahQuinones** [comment]
/lgtm 

Testing 

```
kubectl get csinodes  i-06465a7732d4d80df -o yaml -n kube-system                                                                                                                                                ─╯
apiVersion: storage.k8s.io/v1
kind: CSINode
metadata:
  annotations:
    storage.alpha.kubernetes.io/migrated-plugins: kubernetes.io/aws-ebs,kubernetes.io/azure-disk,kubernetes.io/azure-file,kubernetes.io/cinder,kubernetes.io/gce-pd,kubernetes.io/portworx-volume,kubernetes.io/vsphere-volume
  creationTimestamp: "2025-04-09T15:52:06Z"
  name: i-06465a7732d4d80df
  ownerReferences:
  - apiVersion: v1
    kind: Node
    name: i-06465a7732d4d80df
    uid: *removed*
  resourceVersion: "10005"
  uid: *removed
spec:
  drivers:
  - allocatable:
      count: 26
    name: ebs.csi.aws.com
    nodeID: i-06465a7732d4d80df
    topologyKeys:
    - kubernetes.io/os
    - topology.ebs.csi.aws.com/zone
    - topology.kubernetes.io/zone
```

Attached ENI through AWS console

```
kubectl get csinodes  i-06465a7732d4d80df -o yaml -n kube-system                                                                                                                                                ─╯
apiVersion: storage.k8s.io/v1
kind: CSINode
metadata:
  annotations:
    storage.alpha.kubernetes.io/migrated-plugins: kubernetes.io/aws-ebs,kubernetes.io/azure-disk,kubernetes.io/azure-file,kubernetes.io/cinder,kubernetes.io/gce-pd,kubernetes.io/portworx-volume,kubernetes.io/vsphere-volume
  creationTimestamp: "2025-04-09T15:52:06Z"
  name: i-06465a7732d4d80df
  ownerReferences:
  - apiVersion: v1
    kind: Node
    name: i-06465a7732d4d80df
    uid: *removed*
  resourceVersion: "366032"
  uid: *removed*
spec:
  drivers:
  - allocatable:
      count: 25
    name: ebs.csi.aws.com
    nodeID: i-06465a7732d4d80df
    topologyKeys:
    - kubernetes.io/os
    - topology.ebs.csi.aws.com/zone
    - topology.kubernetes.io/zone
```

```
│ I0410 19:34:17.982775       1 node.go:574] "NodeGetInfo: called" args=""                                                                                                                                           │
│ I0410 19:34:17.985782       1 node.go:781] "getVolumesLimit:" instanceType="c5.large"                                                                                                                              │
│ I0410 19:34:17.986023       1 node.go:785] "getVolumesLimit:" availableAttachments=28                                                                                                                              │
│ I0410 19:34:17.986139       1 node.go:790] "getVolumesLimit:" numBlockDevices=0                                                                                                                                    │
│ I0410 19:34:17.986189       1 node.go:792] "getVolumesLimit:" reservedVolumeAttachments=1                                                                                                                          │
│ I0410 19:34:17.986315       1 node.go:806] "getVolumesLimit:" ENIs=2                                                                                                                                               │
│ I0410 19:34:17.986354       1 node.go:808] "getVolumesLimit:" reservedSlots=0                                                                                                                                      │
│ I0410 19:34:17.986395       1 node.go:601] "NodeGetInfo:" maxVolumesPerNode=25
```

Detached Extra ENI 

```
│ I0410 19:35:23.983301       1 node.go:574] "NodeGetInfo: called" args=""                                                                                                                                           │
│ I0410 19:35:23.986562       1 node.go:781] "getVolumesLimit:" instanceType="c5.large"                                                                                                                              │
│ I0410 19:35:23.987433       1 node.go:785] "getVolumesLimit:" availableAttachments=28                                                                                                                              │
│ I0410 19:35:23.988160       1 node.go:790] "getVolumesLimit:" numBlockDevices=0                                                                                                                                    │
│ I0410 19:35:23.988811       1 node.go:792] "getVolumesLimit:" reservedVolumeAttachments=1                                                                                                                          │
│ I0410 19:35:23.989498       1 node.go:806] "getVolumesLimit:" ENIs=1                                                                                                                                               │
│ I0410 19:35:23.990137       1 node.go:808] "getVolumesLimit:" reservedSlots=0                                                                                                                                      │
│ I0410 19:35:23.990742       1 node.go:601] "NodeGetInfo:" maxVolumesPerNode=26                                                                                                                                     
```

```
kubectl get csinodes  i-06465a7732d4d80df -o yaml -n kube-system                                                                                                                                                ─╯
apiVersion: storage.k8s.io/v1
kind: CSINode
metadata:
  annotations:
    storage.alpha.kubernetes.io/migrated-plugins: kubernetes.io/aws-ebs,kubernetes.io/azure-disk,kubernetes.io/azure-file,kubernetes.io/cinder,kubernetes.io/gce-pd,kubernetes.io/portworx-volume,kubernetes.io/vsphere-volume
  creationTimestamp: "2025-04-09T15:52:06Z"
  name: i-06465a7732d4d80df
  ownerReferences:
  - apiVersion: v1
    kind: Node
    name: i-06465a7732d4d80df
    uid: *removed*
  resourceVersion: "367173"
  uid: *removed*
spec:
  drivers:
  - allocatable:
      count: 26
    name: ebs.csi.aws.com
    nodeID: i-06465a7732d4d80df
    topologyKeys:
    - kubernetes.io/os
    - topology.ebs.csi.aws.com/zone
    - topology.kubernetes.io/zone
```

**Review by ElijahQuinones — APPROVED** [review]

**Review by AndrewSirenko — COMMENTED** [review]
Manually tested fallback K8sMetadata behavior when IMDS disabled. Works!

Left a few minor comments

File: pkg/cloud/metadata/metadata.go [hunk 1]
```diff
@@ -69,6 +71,23 @@ func NewMetadataService(cfg MetadataServiceConfig, region string) (MetadataServi
 	return nil, errors.New("IMDS metadata and Kubernetes metadata are both unavailable")
 }
 
+// UpdateMetadata refreshes ENI information.
```
[the remark this answers is not in the document]
Comment (torredil) [reply pkg/cloud/metadata/metadata.go:74]: Done.

File: pkg/cloud/metadata/metadata.go [hunk 2]
```diff
@@ -69,6 +71,23 @@ func NewMetadataService(cfg MetadataServiceConfig, region string) (MetadataServi
...
+		return nil
+	}
+
+	attachedENIs, err := getAttachedENIs(m.MetadataClient)
+	if err != nil {
+		return fmt.Errorf("failed to update ENI count: %w", err)
+	}
+	m.NumAttachedENIs = attachedENIs
+
+	klog.V(7).InfoS("Updated metadata", "numAttachedENIs", m.NumAttachedENIs)
```
[the remark this answers is not in the document]
Comment (torredil) [reply pkg/cloud/metadata/metadata.go:87]: Done.

File: pkg/driver/node.go [hunk 3]
```diff
@@ -573,6 +573,10 @@ func (d *NodeService) NodeGetCapabilities(ctx context.Context, req *csi.NodeGetC
 func (d *NodeService) NodeGetInfo(ctx context.Context, req *csi.NodeGetInfoRequest) (*csi.NodeGetInfoResponse, error) {
 	klog.V(4).InfoS("NodeGetInfo: called", "args", req)
 
+	if err := d.metadata.UpdateMetadata(); err != nil {
+		klog.ErrorS(err, "Failed to update metadata, using cached values")
```
[the remark this answers is not in the document]
Comment (torredil) [reply pkg/driver/node.go:577]: Can you clarify what you mean by an operator being paged from a log message in the container? If the client is available then we expect to be able to refresh the metadata - that operation failing would be considered an error and should be logged as such (its not meant to be an informational log).

File: pkg/cloud/metadata/metadata.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (torredil) [reply pkg/cloud/metadata/metadata.go:74]: @rdpsin I considered this and ultimately decided that it makes more sense to opt for efficiency and remove the extra call since it is unlikely that EC2 will change the existing behavior. As you mentioned, this isn't a one way door, so we can update this code path if the block device map information provided by IMDS is no longer static at a future point in time. Having said that, I'm happy to add the extra call back in if you recommend doing so.

File: pkg/driver/node.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (AndrewSirenko) [reply pkg/driver/node.go:577]: A common pattern many Kubernetes customers partake in is to ingest and alarm on any `ERROR` level logs in containers in critical pods. 

For more context on logs-threshold-alerting see either of the following resources:
- Cloudwatch: https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Alarm-On-Logs.html
- ELK stack: https://www.elastic.co/guide/en/observability/current/logs-threshold-alert.html

---

Typically:
- ERROR logs indicate that critical system functionality is down and needs urgent attention
- WARN level logs signal potential issues that are not urgent/critical errors

Because updating metadata is not critical for ebs-csi-node to function (driver will just use cached data), I think an ERROR severity is too high a severity here (as opposed to a WARN or INFO(4).  

If we want to consider that operation failure an error I believe that incrementing an `_update_metadata_errors_count` metric is a better practice because customers can choose to specifically alarm on this (or ignore it). 

But I'm open to disagree and committing if you feel that this is a critical error.

[the remark this answers is not in the document]
Comment (torredil) [reply pkg/driver/node.go:577]: That is not a common pattern, its an extreme example. I understand that its feasible to create a metric filter that lets you alarm on every single log statement - in any case, that shouldn't stop us from doing the right thing. See the official [documentation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use) / guidance on logging in Kubernetes from `sig-instrumentation`: 

```
klog.ErrorS - Errors should be used to indicate unexpected behaviours in code, like unexpected errors returned by subroutine function calls. Logs generated by ErrorS command may be enhanced with additional debug information (by logging library). Calling ErrorS with nil as error may be acceptable if there is error condition that deserves a stack trace at this origin point.
```

As I stated above, if the client is available then we expect to be able to refresh the metadata - that operation failing would be an **_unexpected_** error, as I understand it. If you feel strongly about this I will move the log statement to `InfoS`.

[the remark this answers is not in the document]
Comment (ConnorJC3) [reply pkg/driver/node.go:577]: Suggest `ErrorS` - IMDS unavailability would be an unexpected issue. As I understand the AWS client should retry the call itself 2 or 3 times, so any extremely momentary issue would be smoothed over.

File: pkg/driver/node_test.go [hunk 4]
```diff
@@ -1960,6 +1960,7 @@ func TestNodeGetInfo(t *testing.T) {
 				m.EXPECT().GetInstanceID().Return("i-1234567890abcdef0")
 				m.EXPECT().GetAvailabilityZone().Return("us-west-2a")
 				m.EXPECT().GetOutpostArn().Return(arn.ARN{})
+				m.EXPECT().UpdateMetadata().Return(nil)
```
Comment (ConnorJC3) [inline pkg/driver/node_test.go:1963]: Should we maybe add a test for if this returns not nil?

File: pkg/cloud/metadata/ec2.go [hunk 5]
```diff
@@ -111,6 +106,7 @@ func EC2MetadataInstanceInfo(svc EC2Metadata, regionFromSession string) (*Metada
 		AvailabilityZone:       doc.AvailabilityZone,
 		NumAttachedENIs:        attachedENIs,
 		NumBlockDeviceMappings: blockDevMappings,
+		MetadataClient:         svc,
```
Comment (ConnorJC3) [inline pkg/cloud/metadata/ec2.go:109]: Maybe consider `EC2MetadataClient` or `IMDSMetadataClient` to make it obvious this is only for that path? Personally I think the long term best approach would be we split these structs anyways, but for now so that it's clear this client isn't used in the k8s path

Comment (torredil) [reply pkg/cloud/metadata/ec2.go]: Done.

File: pkg/driver/node.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (AndrewSirenko) [reply pkg/driver/node.go:577]: Sounds good, resolving.

File: pkg/cloud/metadata/ec2.go (hunk 5 above)
Comment (AndrewSirenko) [reply pkg/cloud/metadata/ec2.go]: +1

And +1 on splitting structs eventually but that should be separate task

File: pkg/driver/node_test.go (hunk 4 above)
Comment (torredil) [reply pkg/driver/node_test.go:1963]: Done.

**Review by ConnorJC3 — COMMENTED** [review]
/lgtm

File: pkg/cloud/metadata/metadata.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rdpsin) [reply pkg/cloud/metadata/metadata.go:74]: I'm fine with keeping it as-is. 

FWIW, I don't think it would affect performance (at least not in any measurable way). 

> unlikely that EC2 will change the existing behavior

Maybe, maybe not. My (non-blocking) concern is that we'll forget about this in the future (if ever EC2 changes its behavior). 

But like I said, happy to disagree and commit. :)
